### PR TITLE
Trigger `OperationTracer` on contexts enter & exit

### LINKED ISSUE #5728: Trace context changes in the EVM
Geth offers a more fine grained tracing capability than Besu. Currently in Besu it is not possible to trace the EVM on a context change (via call, create or selfdestruct). In Geth those methods are defined as:

```
// CaptureEnter is called when EVM enters a new scope (via call, create or selfdestruct).
func (t *opcounter) CaptureEnter(op vm.OpCode, from common.Address, to common.Address, input []byte, gas uint64, value *big.Int) {}

// CaptureExit is called when EVM exits a scope, even if the scope didn't
// execute any code.
func (t *opcounter) CaptureExit(output []byte, gasUsed uint64, err error) {}
```
For consistency I would name those methods `traceContextEnter` and `traceContextExit`.

### PR DESCRIPTION
**Pull request by delehef**
## PR description

Extends the `OperationTracer` interface to be triggered on contexts enter & exit.

@daniellehrner @shemnon 

## Fixed Issue(s)

Fixes #5728

### CODE REVIEW

File: evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java [hunk 1]
```diff
@@ -200,20 +207,12 @@ public void process(final MessageFrame frame, final OperationTracer operationTra
...
-      revert(frame);
-    }
-
-    if (frame.getState() == MessageFrame.State.COMPLETED_SUCCESS) {
-      completedSuccess(frame);
-    }
-
-    if (frame.getState() == MessageFrame.State.COMPLETED_FAILED) {
-      completedFailed(frame);
+    switch (frame.getState()) {
```
Comment (daniellehrner) [inline evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java:210]: AFAIK the `switch` statement is not equivalent to the chain of `if`s. More than one if condition could be triggered, if the state changes during the execution of `exceptionalHalt()`, `revert()`, `completedSuccess()` or `completedFailed()`. I think we need to revert this change.

File: evm/src/main/java/org/hyperledger/besu/evm/tracing/OperationTracer.java [hunk 2]
```diff
@@ -69,7 +69,7 @@ default void traceAccountCreationResult(
    *
    * @param transaction the transaction which will be processed
    */
-  default void traceStartTransaction(final Transaction transaction) {}
+  default void traceStartTransaction(@SuppressWarnings("unused") final Transaction transaction) {}
```
Comment (daniellehrner) [inline evm/src/main/java/org/hyperledger/besu/evm/tracing/OperationTracer.java:72]: I think the `@SuppressWarnings` annotations are not strictly needed here. The CI pipeline should be okay with it not being used, because before it passed for `traceStartTransaction` as well.

**Review by daniellehrner — CHANGES_REQUESTED** [review]

File: evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java (hunk 1 above)
Comment (delehef) [reply evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java:210]: Fair point.

File: evm/src/main/java/org/hyperledger/besu/evm/tracing/OperationTracer.java (hunk 2 above)
Comment (delehef) [reply evm/src/main/java/org/hyperledger/besu/evm/tracing/OperationTracer.java:72]: To be frank I didn't understand why it complained about unused parameters in `traceStartTransaction` but not in `traceEndTransaction`...

File: evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java [hunk 3]
```diff
@@ -192,6 +208,9 @@ public void process(final MessageFrame frame, final OperationTracer operationTra
       codeExecute(frame, operationTracer);
 
       if (frame.getState() == MessageFrame.State.CODE_SUSPENDED) {
+        if (operationTracer != null) {
```
Comment (shemnon) [inline evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java:211]: Suspended does not mean the context is exiting.  Really what it means is that we are starting a new sub-context and the particulars of how this state machine is written we need to restart from the beginning.

This should be removed.

File: evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java [hunk 4]
```diff
@@ -146,7 +152,10 @@ protected void revert(final MessageFrame frame) {
    *
    * @param frame The message frame
    */
-  private void completedSuccess(final MessageFrame frame) {
+  private void completedSuccess(final MessageFrame frame, final OperationTracer operationTracer) {
+    if (operationTracer != null) {
```
Comment (shemnon) [inline evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java:156]: I'd rather see these in the `process` method and these signatures not altered (like was done with the `CODE_SUSPENDED` case).  The `contextEnter` call also occurs there.

File: evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java [hunk 5]
```diff
@@ -158,7 +167,10 @@ private void completedSuccess(final MessageFrame frame) {
    *
    * @param frame The message frame
    */
-  private void completedFailed(final MessageFrame frame) {
+  private void completedFailed(final MessageFrame frame, final OperationTracer operationTracer) {
+    if (operationTracer != null) {
```
Comment (shemnon) [inline evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java:171]: I'd rather see these in the `process` method and these signatures not altered (like was done with the `CODE_SUSPENDED` case).  The `contextEnter` call also occurs there.

File: evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java [hunk 6]
```diff
@@ -124,7 +124,10 @@ private void clearAccumulatedStateBesidesGasAndOutput(final MessageFrame frame)
    *
    * @param frame The message frame
    */
-  private void exceptionalHalt(final MessageFrame frame) {
+  private void exceptionalHalt(final MessageFrame frame, final OperationTracer operationTracer) {
+    if (operationTracer != null) {
+      operationTracer.traceContextExit(frame);
+    }
```
Comment (shemnon) [inline evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java:130]: Note that in line 134 the state is set to `COMPLETED_FAILED` and this will result in context edit called twice

File: evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java [hunk 7]
```diff
@@ -136,7 +139,10 @@ private void exceptionalHalt(final MessageFrame frame) {
    *
    * @param frame The message frame
    */
-  protected void revert(final MessageFrame frame) {
+  protected void revert(final MessageFrame frame, final OperationTracer operationTracer) {
+    if (operationTracer != null) {
+      operationTracer.traceContextExit(frame);
+    }
```
Comment (shemnon) [inline evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java:145]: Note that in line 147 the state is set to `COMPLETED_FAILED` so the context exit will be called twice.

**Review by shemnon — CHANGES_REQUESTED** [review]
This change needs tests.  Either directly such as how `ExtendedOperationTracerTest` tested the code or indirectly like how `traceEndTransaction` is integrated into the testing of the `T8nExecutor`.

File: evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java (hunk 5 above)
Comment (delehef) [reply evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java:171]: In which scenario can the `contextEnter` happen here?

File: evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java [hunk 8]
```diff
@@ -208,31 +198,45 @@ public void process(final MessageFrame frame, final OperationTracer operationTra
...
-        if (operationTracer != null) {
-          operationTracer.traceContextExit(frame);
-        }
         return;
       }
 
       if (frame.getState() == MessageFrame.State.CODE_SUCCESS) {
+        if (operationTracer != null) {
+          operationTracer.traceContextExit(frame);
+        }
```
Comment (shemnon) [inline evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java:207]: This is unneeded.  We should only do this on `COMPLETED_SUCCESS` and `COMPLETED_FAILURE`

File: evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java [hunk 9]
```diff
@@ -208,31 +198,45 @@ public void process(final MessageFrame frame, final OperationTracer operationTra
...
         codeSuccess(frame, operationTracer);
       }
     }
 
     if (frame.getState() == MessageFrame.State.EXCEPTIONAL_HALT) {
-      exceptionalHalt(frame, operationTracer);
+      if (operationTracer != null && frame.getMessageStackDepth() > 0) {
+        operationTracer.traceContextExit(frame);
+        exitTraced = true;
+      }
```
Comment (shemnon) [inline evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java:216]: This is unneeded.  We should only do this on `COMPLETED_SUCCESS` and `COMPLETED_FAILURE`

File: evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java [hunk 10]
```diff
@@ -208,31 +198,45 @@ public void process(final MessageFrame frame, final OperationTracer operationTra
...
+      exceptionalHalt(frame);
     }
 
     if (frame.getState() == MessageFrame.State.REVERT) {
-      revert(frame, operationTracer);
+      if (operationTracer != null && !exitTraced && frame.getMessageStackDepth() > 0) {
+        operationTracer.traceContextEnter(frame);
+        exitTraced = true;
+      }
```
Comment (shemnon) [inline evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java:224]: This is unneeded.  We should only do this on `COMPLETED_SUCCESS` and `COMPLETED_FAILURE`

File: evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java [hunk 11]
```diff
@@ -208,31 +198,45 @@ public void process(final MessageFrame frame, final OperationTracer operationTra
...
+      revert(frame);
     }
 
     if (frame.getState() == MessageFrame.State.COMPLETED_SUCCESS) {
-      completedSuccess(frame, operationTracer);
+      if (operationTracer != null && !exitTraced && frame.getMessageStackDepth() > 0) {
+        operationTracer.traceContextEnter(frame);
+        exitTraced = true;
```
Comment (shemnon) [inline evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java:231]: We don't need to track `exitTraced`

File: evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java [hunk 12]
```diff
@@ -196,7 +184,9 @@ private void codeExecute(final MessageFrame frame, final OperationTracer operati
    * @param operationTracer the operation tracer
    */
   public void process(final MessageFrame frame, final OperationTracer operationTracer) {
-    if (operationTracer != null) {
+    boolean exitTraced = false;
```
Comment (shemnon) [inline evm/src/main/java/org/hyperledger/besu/evm/processor/AbstractMessageProcessor.java:187]: We don't need to track `exitTraced`

**Review by daniellehrner — APPROVED** [review]

**Comment by shemnon** [comment]
I was hoping for a test more like https://github.com/hyperledger/besu/blob/532797fc1725d3264ca51d083cb2b8a89123f034/evm/src/test/java/org/hyperledger/besu/evm/StandardJsonTracerTest.java#L33, where there are  no mocks.  A custom tracer that outputs the tracing results into a data structure or text would then be verified.

With mocks-only testing I just don't feel it proves that it's not behaving in surprising ways.

**Comment by daniellehrner** [comment]
@shemnon I added another test without mocks, based on EIP-3155. Could you please review again?

**Review by shemnon — APPROVED** [review]
